include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h

```diff
@@ -1665,125 +1535,86 @@ struct __leaf_sort_kernel
 
-        _PRINT_INFO_IN_DEBUG_MODE(__exec);
-
-        // __leaf: size of a block to sort using algorithm suitable for small sequences
-        // __optimal_chunk: best size of a block to merge duiring a step of the merge sort algorithm
-        // The coefficients were found experimentally
-        _Size __leaf = 4;
-        _Size __optimal_chunk = 4;
-        if (__exec.queue().get_device().is_cpu())
-        {
-            __leaf = 16;
-            __optimal_chunk = 32;
-        }
-        // Assume powers of 2
-        assert((__leaf & (__leaf - 1)) == 0);
-        assert((__optimal_chunk & (__optimal_chunk - 1)) == 0);
-
-        const _Size __leaf_steps = ((__n - 1) / __leaf) + 1;
+        const bool __is_cpu = __exec.queue().get_device().is_cpu();
+        const ::std::uint32_t __leaf = __is_cpu ? 16 : 2;
+        _Size __steps = oneapi::dpl::__internal::__dpl_ceiling_div(__n, __leaf);
 
         // 1. Perform sorting of the leaves of the merge sort tree
         sycl::event __event1 = __exec.queue().submit([&](sycl::handler& __cgh) {
             oneapi::dpl::__ranges::__require_access(__cgh, __rng);
-            __cgh.parallel_for<_LeafSortName...>(sycl::range</*dim=*/1>(__leaf_steps),
-                                                 [=](sycl::item</*dim=*/1> __item_id) {
-                                                     const _Size __idx = __item_id.get_linear_id() * __leaf;
-                                                     const _Size __start = __idx;
-                                                     const _Size __end = sycl::min(__start + __leaf, __n);
-                                                     __leaf_sort_kernel()(__rng, __start, __end, __comp);
-                                                 });
+            __cgh.parallel_for<_LeafSortName...>(sycl::range</*dim=*/1>(__steps), [=](sycl::item</*dim=*/1> __item_id)
+            {
+                const _IdType __i_elem = __item_id.get_linear_id() * __leaf;
+                __leaf_sort_kernel()(__rng, __i_elem, std::min<_IdType>(__i_elem + __leaf, __n), __comp);
+            });
         });
 
-        _Size __sorted = __leaf;
-        // Chunk size cannot be bigger than size of a sorted sequence
-        _Size __chunk = ::std::min(__leaf, __optimal_chunk);
-
+        // 2. Merge sorting
         oneapi::dpl::__par_backend_hetero::__internal::__buffer<_Policy, _Tp> __temp_buf(__exec, __n);
         auto __temp = __temp_buf.get_buffer();
         bool __data_in_temp = false;
+        ::std::uint32_t __n_sorted = __leaf;
+        const ::std::uint32_t __chunk = __is_cpu ? 32 : 4;
+        __steps = oneapi::dpl::__internal::__dpl_ceiling_div(__n, __chunk);
 
-        // 2. Perform merge sorting
-        // TODO: try to presort sequences with the same approach using local memory
-        while (__sorted < __n)
+        const ::std::size_t __n_power2 = oneapi::dpl::__internal::__dpl_bit_ceil(__n);
+        const ::std::size_t n_iter = std::log2(__n_power2) - 1;
```

Comment:
I believe the number of iterations needs to depend on the size of the leaf (this may also explain why it did not yield performance improvement in earlier testing...).

Reply:
```suggestion
        const ::std::int64_t n_iter = ::std::log2(__n_power2) - ::std::log2(__leaf);
```
I think this needs to be signed in case the `__leaf` is bigger than `__n_power2`.

Reply:
Yes, "`- ::std::log2(__leaf);`".
I've already written like that.

Reply:
I do believe `n_iter` needs to be signed though unless we filter out very small cases higher up the stack.